### ISSUE TITLE
Allow disabling the sorting timeout

### ISSUE THREAD

**Issue by afdw** [htop-dev/htop#1410]
Currently, sorting only takes effect after some time has passed: https://github.com/htop-dev/htop/blob/4feac8e72f76c7d50ac4e61689645796cd5fdc31/ScreenManager.c#L240-L241

Also, the timeout is reset every time an unknown key is pressed: https://github.com/htop-dev/htop/blob/4feac8e72f76c7d50ac4e61689645796cd5fdc31/ScreenManager.c#L420

Meaning, if an unbound key such as `o` is pressed, sorting is paused for a few seconds. For me, this also happens when the window of a terminal emulator where htop is running gains of looses focus. So I have to wait that time after switching to the htop window to see an up-to-date information about what is consuming CPU, or find another way to trigger sorting to start updating, such as reversing the sorting direction twice or pressing `Z` twice.

As far as I understand, the purpose of the sorting timeout is to give the user time to read the process list before it starts flashing and changing order. But I do not see why the timeout is always reset in cases described above.

I find all that a bit confusing and I would prefer the sorting to be always consistently applied without any timeouts. Would it be possible to add an option to complete disable the sorting timeout? Would a simple on/off option suffice?

**Comment by Explorer09** [comment]
"Disabling" the sorting timeout is not an easy task as it requires adding a real-time algorithm to insert and sort the entries at the same time. (This is computer science knowledge.)

Is there any real reason you want this real-time update and sorting? I thought a 500 ms sorting period should be enough for humans to not observe any sorting delay.

**Comment by afdw** [comment]
> "Disabling" the sorting timeout is not an easy task as it requires adding a real-time algorithm to insert and sort the entries at the same time. (This is computer science knowledge.)

It seems that I did not guess correctly the purpose of the timeout. Can you explain it? Why would it be an issue for `sortTimeout` to always be `0`, what would that break? And why a special algorithm would be needed?

> Is there any real reason you want this real-time update and sorting? I thought a 500 ms sorting period should be enough for humans to not observe any sorting delay.

For me currently it is more than 3 seconds. But even then, I do think 0.5 seconds would be noticeable: this is enough time for a human to look at the first row in the table and get to a wrong conclusion about which process consumes the most resources, if the table is not properly updated.

And again, why is there need to reset this timeout every time an unknown key is pressed or the htop window changes focus?

**Comment by Explorer09** [comment]
> It seems that I did not guess correctly the purpose of the timeout. Can you explain it? Why would it be an issue for `sortTimeout` to always be `0`, what would that break? And why a special algorithm would be needed?

Think of holding the "F5" or Refresh button on your web browser when you are surfing the web. Every reloading action of the web page would takes CPU time, and if you keep the browser refreshing the page, you won't be able to read or use the page at all.

I think this analogy works because htop doesn't update the process list through any asynchronous mechanism. htop won't be notified if any new process is created or any process dies. So what htop can do is to poll the `procfs` at a set interval. That's the reason for the timeout.

**Comment by afdw** [comment]
> I think this analogy works because htop doesn't update the process list through any asynchronous mechanism. htop won't be notified if any new process is created or any process dies. So what htop can do is to poll the `procfs` at a set interval. That's the reason for the timeout.

This shows why the process list is only updated periodically, but that does not explain the reason behind the sorting timeout.

More precisely, what would go wrong if this line: https://github.com/htop-dev/htop/blob/4feac8e72f76c7d50ac4e61689645796cd5fdc31/ScreenManager.c#L241

Would be written instead as follows?
```c
int resetSortTimeout = 0;
```

What is the purpose of the `sortTimeout` variables in this file?

**Comment by afdw** [comment]
I have recorded a video demonstrating the issue. Here I launch a calculator and htop with sorting by the `CPU%` column. As you can see, when I switch between the calculator and htop windows, the sorting by the CPU usage stops working for a few seconds before kicking back in. This is the timeout I am talking about.

https://github.com/htop-dev/htop/assets/7559560/51021b8b-e627-462e-aed1-eb713e36fa33

**Comment by BenBE** [comment]
The reason the process list isn't updated after key presses immediately is due to UX reasons. Think of a busy system where you want to kill a certain process. If you navigate the list and every time you hit a key the list refreshed you'd end up with quite a busy, ever-changing list of processes where the one you are trying to kill or otherwise interact with would keep jumping around. By disabling the sorting after a key press you allow the user to focus on what they want to do: navigate to the target process and hit the intended action key.

The only problem I see with that right now, is the fact, that this sorting inhibitor is also triggered if no actionable key for htop was pressed. And that's the only direction I'd accept taking any change in behaviour on this front.

**Comment by BenBE** [comment]
@Explorer09 I know. And the cause is with how key presses are handled: Currently each keypress detected in the main loop resets the timeout, when only key presses that we actually handle should do the reset. Like I said in my previous comment … ;-)

**Comment by fasterit** [comment]
@afdw Can you please record what (simulated) keystrokes your desktop environment sends to the terminal window when you click it?

Run `xev -event keyboard -event button` and click the top bar of the xev window to activate it.
Please send the event(s) that you receive that way.

**Comment by afdw** [comment]
> @afdw Can you please record what (simulated) keystrokes your desktop environment sends to the terminal window when you click it?

I think the more important question is not what the X11 window of the terminal emulator receives, but what htop gets from it on the stdin.

To determine that, I ran htop while saving the stdin contents to a file, immediately quit htop using `q` and then dumped the file:
```
user@notebook:/tmp/tmp.w0RnPLtqMz$ htop < <(tee out)
user@notebook:/tmp/tmp.w0RnPLtqMz$ hexdump -C out
00000000  1b 5b 49 71                                       |.[Iq|
00000004
```

So, we can tell that `1b 5b 49` is printed on startup (and thus introduces sorting delay), and `71` is me pressing `q`.

Now, I did the same thing, but this time instead of immediately quitting, I did what was recorded in the video above — switched between the htop window and another one:
```
user@notebook:/tmp/tmp.w0RnPLtqMz$ htop < <(tee out)
user@notebook:/tmp/tmp.w0RnPLtqMz$ hexdump -C out
00000000  1b 5b 49 1b 5b 4f 1b 5b  49 1b 5b 4f 1b 5b 49 1b  |.[I.[O.[I.[O.[I.|
00000010  5b 4f 1b 5b 49 71                                 |[O.[Iq|
00000016
```

We can see that the sequence `1b 5b 4f` and `1b 5b 49` are received when htop gains or looses focus.

This behavior is documented here: https://invisible-island.net/xterm/ctlseqs/ctlseqs.html#h3-FocusIn_FocusOut. It is present (at lest) in `xterm` and `xfce4-terminal`.

A link that might be relevant about how this can interact with ncurses: https://stackoverflow.com/questions/54472426/ncurses-detect-when-mouse-leaves-window.
